hackjohn.py

```diff
@@ -9,175 +9,522 @@
 According to the reservations office,
 the table is usually updated around 11 AM pacific time
 and spaces are usually snatched within ten minutes.
-Call the reservation number if there's availability at 209-372-0740.
+Fill out the online permit request form if there are available permits:
+https://yosemite.org/yosemite-wilderness-permit-request-form/
 """
 
 import pathlib
-import re
-
 import requests
-import pandas
-from pkg_resources import parse_version
-
-# Minimum number of available spaces
-spaces = 2
+import json
+from datetime import datetime, timedelta
+import pytz
+from twocaptcha import TwoCaptcha
+from typing import Tuple, Union, List
+import warnings
+import pickle
+from tenacity import (
+    retry,
+    stop_after_attempt,
+    wait_fixed,
+    retry_if_exception_type
+)
 
-# Comment out trailheads you'd like to start from
-exclude = [
-    # "HI → LYV",
-    # "HI → Sunrise/Merced Lakes (Pass through)",
-    # "Glacier Point → LYV",
+# some variables are set in config.py -- not intended to be modified
+import config
+
+
+######################
+# Set these parameters
+######################
+CAPTCHA_API_KEY = "your-api-key"  # replace with your 2Captcha API key
+MIN_SPACES = 1  # minimum number of available spaces
+START_DATE = "2021-06-15"  # earliest date you would like to start your hike
+END_DATE = "2021-09-30"    # latest date you would like to start your hike
+NOTIFY_IF_NO_PERMITS = True  # send daily notification even if no permits are found?
+OUTPUT_TIME_ZONE = "US/Pacific"  # what time zone to show in notifications (must be in pytz.all_timezones)
+
+# Comment out trailheads you'd like to start from (don't change the trailhead names)
+EXCLUDE_TRAILHEADS = [
+    # "Happy Isles->Little Yosemite Valley",
+    # "Happy Isles->Sunrise/Merced Lake (pass through)",
+    # "Glacier Point->Little Yosemite Valley",
     # "Sunrise Lakes",
     # "Lyell Canyon",
 ]
```

Comment:
This is what I intended when I set it up that way:
* `hackjohn.py` is the only file the user will need to edit before running. It contains all user API keys; user preferences around dates, trailheads, etc.; and notification settings.
* `config.py` contains constants that the user doesn't need to see or know about. We might need to change those values if the website or permit process changes in the future, so it's kind of helpful from a development to have those values in a separate file rather than hardcoding them into hackjohn.py.

I'm open to restructuring though. We could either move all of parameters to config.py, or we could add another file for those parameters (for example, user_config.py for parameters the user should edit and system_config.py for the values that shouldn't be changed). Or we could even go back to having everything included in hackjohn.py and not import any other files.

I don't really feel strongly any which way. Do you have any preference?

Reply:
My worry with the parameters in hackjohn.py is that this file might be a bit more daunting to users compared to a more simple config file. So in that sense I like the idea of:

1. user_config.py / system_config.py
2. config.py with both but comments separating the two sections.

Up to you. Also can be done as a subsequent PR to this.

Reply:
I'll keep the single config.py file and separate the two sections with comments. Should be quick so I'll do it in this PR.